### ISSUE TITLE
Feature: Send Image/Attachments... Custom slots?

### ISSUE THREAD

**Issue by martinandersen3d** [MatheusrdSantos/vue-quick-chat#16]
In your project I have to send pictures in the chat. I do not want to create a fork, but rather improve this project.

Thought it would probably be best to discuss how it can be implemented here first.

My first thought is that we can have two custom slots. One for pictures and one for attachments.

--------------------------
Suggested solution:

step 1 - Register the input:
![image](https://user-images.githubusercontent.com/19984581/68960135-d74d1800-07cf-11ea-9340-f4f9aae6af55.jpg)

step 2 - the programmer does whatever he wants with the callback.

3. Now it needs to be displayed in the chat..

Here is some possible suggestions/solutions:

- There are custom slots for different file types
- or no slots, but the json define how the "chat buble" will look like.
- Can you think of more options?

**What are your thoughts on that?** :-)

**Comment by MatheusrdSantos** [comment]
Hi, @martinandersen3d 
It's in my plans to implement such functionallities. However I'm quite busy, so I can't implement it yet (my plans are implement new functionalities in the end of this month).
In my opnion, your suggestions are good and I'll follow it.

About step 3, I gess you can have two cases (two slot types):
- images (the chat displays a preview of the uploaded or received image)
- other file type (the chat displays an element with the filename and the file extension)
(I think the most simple and generic solution is better, because we can implement new functionlities after)

steps 1 and 2 are ok.
Remember that probably the message object will have three new fields: messageType ('image', 'file' or 'text'), contentUrl (if the message type is 'file' or 'image') and maybe contentSize (the file/image size in kb/mb)

Thaks for your feedback.

**Comment by martinandersen3d** [comment]
I completely agree with you. 

If we implementing the _messageType_  a fourth option could be "html". Then it is very very customizable.

My client is also talking about video chat in the future. Having the ability to insert raw html, would make it possible to do any custom styling and flexibility. Good for any features, that the client will come up with :D

So to add to your comment:

- messageType ('image', 'file', 'text', **'html'**)
- contentUrl (if the message type is 'file' or 'image') and maybe
- contentSize (the file/image size in kb/mb)
- _contentHtml_ (for custom html)

**Comment by martinandersen3d** [comment]
:on-image-upload
:on-file-upload

**Comment by Tofandel** [comment]
@martinandersen3d Raw html would make it susceptible to XSS attacks, sure it's up to the server/dev to prevent this, but I'd rather not have accidental vulnerabilities

I would also only add an option :on-file-upload type "function" if it's defined then display the attachment button (which should be slotted as all other icons)

And another option for file types accepted, :file-types which could be defaulted to images (.jpg,.jpeg,.png,.gif)

It's a straightforward dev (shouldn't take more than 1hour) so you can also do a PR

**Comment by martinandersen3d** [comment]
Sounds good.

I do also agree on the raw html.

**Links:**
I think if we make it possible to send links in the chat, then I would be able to forward any "video" chat features, via a link to another router in the application.

Maybe we could add it like:
messageType ('image', 'file', 'text')
contentUrl (if the message type is 'file' or 'image') and maybe
contentSize (the file/image size in kb/mb)
LinkUrl: ...
LinkText: ...
LinkImg:.. I don't know if there could/should be a image in the link, 
LinkId  = Html Element ID
LinkClass  = Html Element Class

I'm not sure what is the best way to do this, in a flexible way.. 

What do you think?

**Comment by Tofandel** [comment]
Links is another story. That would be a valid use case for using v-html however the message should be entirely html escaped first

Then it could be parsed with a regex and if you find a link add an <a> tag for the link

When can also give control to users who knows what they are doing to v-html via the message type 'html'

**Comment by martinandersen3d** [comment]
I just came to mind that vue has what is called a **dynamic component**. Vue has an "is" tag that it can convert to an HTML element.

Here is the Vue code (html element type is defined in the **"IS="**):
```
<div is="a" href = "www .." class = "foo" style = "color: red">
   anything inside component
</ div>
```

It will be converted to HTML:
```
<a href="www.." class="foo" style="color:red">
   anything inside component
</a>
```
Vue does not allow "script" inside the tag, so safety is good.

I think this would be a good, secure and flexible solution.

You can see examples, and documentation here:

- https://codepen.io/martinandersen3d/pen/yLLrNYb
- https://stackoverflow.com/questions/41748934/how-to-create-dynamic-tag-based-on-props-with-vue-2
- https://vuejs.org/v2/guide/components.html#Dynamic-Components


Example:
![image](https://user-images.githubusercontent.com/19984581/69386171-142e7880-0cc2-11ea-9487-44324e92b5ec.png)

**Comment by Tofandel** [comment]
This will not be helpful, Vue is not interpolated in v-html, and it's anyway the same as writtng an a tag, this feature is only useful when you need to change the vue component displayed based on some logic

**Comment by Tofandel** [comment]
I know how to display links without v-html

Instead of escaping it, just explode the text based on regex, parts that match url can be rendered inside an a, the rest in moustache

**Comment by martinandersen3d** [comment]
I think we misunderstand each other :D

I 100% agree that we can/should use a regex to make links, based on text that the user writes. That's not what I'm trying to explain :D : D

What I'm trying to do is:
To create a solution that allows the external programmer to use the component to create the html messages to suit their needs. So make it as flexible as possible and still maintain the security that you talk about :-)

By using dynamic components, we can solve all of the above problems (- minus the regex), as well as enable other programmer to customize every chat messages look and feel, according to their own needs.

Here's what I'm trying to explain:
![chat](https://user-images.githubusercontent.com/19984581/69426331-5a6cf180-0d2d-11ea-8a21-5ea3fa1c9269.jpg)

Now, the solution is not "perfect", since there is edge-cases that it will not solve, but there is a lot of stuff that it CAN do. 

Does it make sense? :D

**Comment by Tofandel** [comment]
Okay I see what you're saying but this poses the problem I mentioned about the server having power over the tags used when this power should be left to vue alone via scoped slots, in the scoped slot you can check the message type and display absolutely any component you want rather than a runtime tag, so it's also more flexible

**Comment by martinandersen3d** [comment]
I actually understand what you are trying to do now :D 

I like it, its genius :D

So from the messageType, contentUrl, contentSize, external developers can create an unlimited numbers of different slots, where we can loop through and have different components with different properties. Thats really nice :D

**Comment by Tofandel** [comment]
Yes basically but not just those properties, since the entire message will be passed as a parameter to the scoped slot, you can use any property you want like this
```
<template #message="tpl">
    <div v-if="tpl.message.type == 'text'">{{tpl.message.content}}</div>
    <a v-elseif="tpl.message.type == 'url'">{{tpl.message.url}}</a>
</template>
```

**Comment by Tofandel** [comment]
If it's urgent, you can do a PR and I'll review it, else I can do it later this week

I would change the name of the component MessageDisplay to MessagesList for clarity and add a component MessageInner (or MessageDisplay) which will only contain the display markup
(It's this part)
```html
            <div class="message-text"
                 :style="{background: !message.myself?colors.message.others.bg: colors.message.myself.bg, color: !message.myself?colors.message.others.text: colors.message.myself.text}">
                <p v-if="!message.myself" class="message-username">{{getParticipantById(message.participantId).name}}</p>
                <p v-else class="message-username">{{myself.name}}</p>
                <p>{{message.content}}</p>
            </div>
            <div class="message-timestamp" :style="{'justify-content': message.myself?'flex-end':'baseline'}">
                {{message.timestamp.format('LT')}}
                <v-icon v-if="asyncMode && message.uploaded" name="check" base-class="icon-sent"/>
                <div v-else-if="asyncMode" class="message-loading"></div>
            </div>
```

**Comment by Tehsseen** [comment]
hi guys is anyone have the updated version with image upload feature. thanks

**Comment by gabrieljbaker** [comment]
Curious about this image upload feature as well!

**Comment by martinandersen3d** [comment]
Anybody working on this? Else I will look into it, this month.. Anybody want to join? :-)

**Comment by MatheusrdSantos** [comment]
@martinandersen3d 
I guess nobody is working on this feature. But I'll implement a way of display images on chat, as I said here #22.

**Comment by martinandersen3d** [comment]
Thanks ;-)

**Comment by vonec** [comment]
+1 .. it would be great if we can have images + files ... i will try to chip in ... been a while ... happy new year @MatheusrdSantos

**Comment by vonec** [comment]
> I completely agree with you.
> 
> If we implementing the _messageType_ a fourth option could be "html". Then it is very very customizable.
> 
> My client is also talking about video chat in the future. Having the ability to insert raw html, would make it possible to do any custom styling and flexibility. Good for any features, that the client will come up with :D
> 
> So to add to your comment:
> 
> * messageType ('image', 'file', 'text', **'html'**)
> * contentUrl (if the message type is 'file' or 'image') and maybe
> * contentSize (the file/image size in kb/mb)
> * **_contentHtml_** (for custom html)
> 
> The challenge is to create a extend-ability that serve many different purposes.. I can't tell if it should be slots or json, but with the _Image_ and _attachment_, its really good start.

if the html part is only about video embeds we can support oembed / graph data much like facebook and twitter but i thinks its a long way to go

**Comment by vonec** [comment]
:send-images="true" is enabling the image attatchment but 

client.js:97 Error: Cannot linkify undefined - Invalid DOM Node type
    at linkifyElementHelper (vue-quick-chat.common.js:679)
    at linkifyElement (vue-quick-chat.common.js:742)
    at VueComponent.mounted (vue-quick-chat.common.js:16295)

when we try to attatch an image ... am i missing something

**Comment by vonec** [comment]
https://github.com/MatheusrdSantos/vue-quick-chat/pull/51 waiting for it get merged hope that would solve a small problem
